### PR TITLE
DEPR: Disallow dtype inference when setting Index into DataFrame

### PR DESCRIPTION
**Pull request by phofl**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

cc @jbrockmendel I think we chatted about this. Series keeps the dtype, so this is mostly for consistency and to make our lives a little easier

### CODE REVIEW

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @phofl

File: pandas/core/frame.py [hunk 1]
```diff
@@ -5219,7 +5219,21 @@ def _sanitize_column(self, value) -> tuple[ArrayLike, BlockValuesRefs | None]:
...
+        arr = sanitize_array(value, self.index, copy=True, allow_2d=True)
+        if (
+            isinstance(value, Index)
+            and value.dtype == "object"
+            and arr.dtype != value.dtype
+        ):  #
+            # TODO: Remove kludge in sanitize_array for string mode when enforcing
+            # this deprecation
+            warnings.warn(
+                "Setting an Index with object dtype into a DataFrame will no longer "
```
Comment (jbrockmendel) [inline pandas/core/frame.py:5231]: is "will no longer" clear enough that this refers to a future version/deprecation?

File: pandas/core/frame.py [hunk 2]
```diff
@@ -5219,7 +5219,21 @@ def _sanitize_column(self, value) -> tuple[ArrayLike, BlockValuesRefs | None]:
 
         if is_list_like(value):
             com.require_length_match(value, self.index)
-        return sanitize_array(value, self.index, copy=True, allow_2d=True), None
+        arr = sanitize_array(value, self.index, copy=True, allow_2d=True)
+        if (
+            isinstance(value, Index)
```
Comment (jbrockmendel) [inline pandas/core/frame.py:5224]: might be more performant to do the Index check before the sanitize_array?

Comment (phofl) [reply pandas/core/frame.py:5224]: Why? We have to sanitise anyway?

File: pandas/core/frame.py (hunk 1 above)
Comment (phofl) [reply pandas/core/frame.py:5231]: good point, I'll put up a PR to clarify

File: pandas/core/frame.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/frame.py:5224]: if you know you have an Index (and no dtype), sanitize_array boils down to:

```
if len(value) != len(index): raise ...
return value._values
```

so a lot of sanitize_array may be made unnecessary in this case

Comment (phofl) [reply pandas/core/frame.py:5224]: That's what I want in the future, but we go through ``maybe_infer_to_datetimelike`` at the moment which changes the dtype, so we can't shortcut this before we enforce the deprecation

**Pull request merged**